tiledb/sm/query/readers/attribute_order_validator.h

```diff
@@ -0,0 +1,581 @@
+      // fragment's max, we can validate. Otherwise we'll need to load
+      // the tile.
+      const IndexType* non_empty_domain2 =
+          static_cast<const IndexType*>(non_empty_domains[f2]);
+      if (min_tile_aligned || min - 1 == non_empty_domain2[1]) {
+        min_validated_[f] = true;
+
+        // Check the order.
+        if (increasing_data) {
+          // Increasing data: The first value on f is the minimum on f. Check
+          // that it is greater than the last (maximum) value in the proceeding
+          // tile on f2.
+          auto value = fragment_metadata[f]->get_tile_min_as<AttributeType>(
+              attribute_name_, 0);
+
+          auto value_previous =
+              fragment_metadata[f2]->get_tile_max_as<AttributeType>(
+                  attribute_name_, f2_tile_idx);
+
+          if (value_previous >= value) {
+            throw AttributeOrderValidatorStatusException(
+                "Attribute out of order");
+          }
+        } else {
+          // Decreasing data: The first value on f is the maximum of f. Check
+          // that is is less than the last (minimum) value in the proceeding
+          // tile on f2.
+          auto value = fragment_metadata[f]->get_tile_max_as<AttributeType>(
+              attribute_name_, 0);
+
+          auto value_previous =
+              fragment_metadata[f2]->get_tile_min_as<AttributeType>(
+                  attribute_name_, f2_tile_idx);
+
+          if (value_previous <= value) {
+            throw AttributeOrderValidatorStatusException(
+                "Attribute out of order");
+          }
+        }
+      } else {
+        // Add the tile to the list of tiles to load.
+        add_tile_to_load(f, true, f2, f2_tile_idx, schema);
+      }
+    }
+
+    if (!max_validated_[f]) {
+      // Get the fragment number to compare against.
+      if (!fragment_to_compare_against_[f].second.has_value()) {
+        throw std::logic_error(
+            "Should know fragment value to compare for non-validated bound.");
+      }
+      auto f2 = fragment_to_compare_against_[f].second.value();
+
+      // Get the max index and max tile index. See if the max is tile aligned.
+      auto max = non_empty_domain[1];
+      auto max_tile_idx = fragment_metadata[f]->tile_num() - 1;
+      bool max_tile_aligned = max + 1 == index_dim->round_to_tile<IndexType>(
+                                             max + 1, dim_dom[0], tile_extent);
+
+      // Get the tile index of the tile in f2 immediately following f (with
+      // respect to tile index in f2). If f is tile aligned we need to add an
+      // additional value.
+      uint64_t f2_tile_idx = max_tile_idx + frag_first_array_tile_idx[f] -
+                             frag_first_array_tile_idx[f2] + max_tile_aligned;
+
+      // If we are tile aligned or the max is right next to the other fragment's
+      // min, we can validate. Otherwise we'll need to load the tile.
```

Comment:
```suggestion
      // If we are tile aligned or non-overlapping (the max is right next to
      // the other fragment's min), we can validate. Otherwise, we'll need to 
      // load the tile.
```